File: Chandra Kiran/12-09-2022/3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function (nums) {
    let result = [];
    nums.sort((a, b) => a - b);
    for (let i = 0; i < nums.length; i++) {
        if (i > 0 && nums[i] === nums[i - 1]) continue;
        let left = i + 1;
        let right = nums.length - 1;
        while (left < right) {
            let sum = nums[i] + nums[left] + nums[right];
            if (sum === 0) {
                result.push([nums[i], nums[left], nums[right]]);
                left++;
                right--;
                while (left < right && nums[left] === nums[left - 1]) left++;
                while (left < right && nums[right] === nums[right + 1]) right--;
            } else if (sum < 0) {
                left++;
            } else {
                right--;
            }
        }
    }
    return result;
};




console.log(threeSum([1, 5, 5, 2, -10, 1]))
console.log(threeSum([-2, 0, 0, 2, 2]))

File: Karthik/12-9_to_13-9/3sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        triplets=[]
        for i,val in enumerate(nums[0:-2]):
            if val>0:
                break
            if val==nums[i-1] and i>0:
                continue
            left=i+1
            right=len(nums)-1
            while(left<right):
                sumof=nums[left]+val+nums[right]
                if sumof==0:
                    triplets.append([nums[left],val,nums[right]])
                if sumof>0:
                    right-=1
                else:
                    left+=1
                    while nums[left]==nums[left-1] and left<right:
                        left+=1
        return triplets
            